### PR TITLE
fix: setProviderAndWait must throw

### PR DESCRIPTION
**Pull request by toddbaert**
As per [1.1.2.4](https://openfeature.dev/specification/sections/flag-evaluation#requirement-1124), the blocking provider setter must throw if init fails, so that application integrators can easily understand if their provider is working without attaching event handlers.

We implemented the wait functionality of this method, but we swallow the exception.

This PR fixes that.

### CODE REVIEW

File: src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java [hunk 1]
```diff
@@ -131,14 +131,14 @@ public void setProvider(String clientName, FeatureProvider provider) {
     /**
      * Set the default provider and wait for initialization to finish.
      */
-    public void setProviderAndWait(FeatureProvider provider) {
+    public void setProviderAndWait(FeatureProvider provider) throws RuntimeException {
```
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java:134]: `RuntimeException` is unchecked, so this is non-breaking.

OpenFeature Errors are all subclasses unchecked, so this is no problem for providers who want to throw OpenFeature errors.

File: src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java [hunk 2]
```diff
@@ -131,14 +131,14 @@ public void setProvider(String clientName, FeatureProvider provider) {
...
         try (AutoCloseableLock __ = lock.writeLockAutoCloseable()) {
             providerRepository.setProvider(
                     provider,
                     this::attachEventProvider,
                     this::emitReady,
                     this::detachEventProvider,
-                    this::emitError,
+                    this::emitErrorAndThrow,
```
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java:141]: This is the main change - I pass a new delegate which not only emits the events, but throws.

File: src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java [hunk 3]
```diff
@@ -179,9 +179,14 @@ private void detachEventProvider(FeatureProvider provider) {
...
-    private void emitError(FeatureProvider provider, String message) {
+    private void emitError(FeatureProvider provider, Exception exception) {
         runHandlersForProvider(provider, ProviderEvent.PROVIDER_ERROR,
-                ProviderEventDetails.builder().message(message).build());
+                ProviderEventDetails.builder().message(exception.getMessage()).build());
+    }
+
+    private void emitErrorAndThrow(FeatureProvider provider, RuntimeException exception) throws RuntimeException {
+        this.emitError(provider, exception);
+        throw exception;
```
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java:189]: New delegate.

File: src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java [hunk 4]
```diff
@@ -179,9 +180,14 @@ private void detachEventProvider(FeatureProvider provider) {
         }
     }
 
-    private void emitError(FeatureProvider provider, String message) {
+    private void emitError(FeatureProvider provider, OpenFeatureError exception) {
         runHandlersForProvider(provider, ProviderEvent.PROVIDER_ERROR,
-                ProviderEventDetails.builder().message(message).build());
+                ProviderEventDetails.builder().message(exception.getMessage()).build());
```
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java:185]: New delegate with emits errors but also throws.

File: src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java [hunk 5]
```diff
@@ -179,9 +180,14 @@ private void detachEventProvider(FeatureProvider provider) {
...
+    }
+
+    private void emitErrorAndThrow(FeatureProvider provider, OpenFeatureError exception) throws OpenFeatureError {
```
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java:188]: New delegate with emits errors but also throws.

**Review by beeme1mr — APPROVED** [review]

**Review by Kavindu-Dodan — APPROVED** [review]

**Pull request merged**